File: Sliding Window/Java/MinSizeSubArraySum.java
/*
Given an array of n positive integers and a positive integer s, 
find the minimal length of a contiguous subarray of which the sum ≥ s. 
If there isn't one, return 0 instead.

Example: 

Input: s = 7, nums = [2,3,1,2,4,3]
Output: 2
Explanation: the subarray [4,3] has the minimal length under the problem constraint.

Follow up:
If you have figured out the O(n) solution, 
try coding another solution of which the time complexity is O(n log n). 

https://leetcode.com/problems/minimum-size-subarray-sum/


*/


public class MinSizeSubArraySum {

    public static int minSubArrayLen(int s, int[] nums){

        int sum = 0;
        int end = 0;
        int beg = 0;
        int minWindowLen = Integer.MAX_VALUE;

        while (end < nums.length){
            sum += nums[end];

            while(sum >= s){
                minWindowLen = Math.min(minWindowLen, end - beg + 1);
                sum -= nums[beg];
                beg++;
            }
            end++;
        }


        return minWindowLen == Integer.MAX_VALUE ? 0 : minWindowLen;

    }

    public static void main (String[] args){


        int s = 7;
        int[] nums = {2,3,1,2,4,3};
        int result = minSubArrayLen(s,nums);

        System.out.println(result);






    }



    
}

File: Sliding Window/Python/MinSizeSubArraySum.py
"""
Given an array of positive integers nums and a positive integer target, return the minimal length of a contiguous subarray [numsl, numsl+1, ..., numsr-1, numsr] 
of which the sum is greater than or equal to target. If there is no such subarray, return 0 instead.
"""



class Solution:
    def minSubArrayLen(self, target, nums):
        for i in range(1,len(nums)+1):
            for j in range(len(nums)):
                arr = nums[j:j+i]
                if(len(arr)==i):
                    if(sum(arr)>=target):
                        return(len(arr))
        return 0



if __name__=="__main__":
    tar = 15
    inp = [1,2,3,4,5]
    print(Solution().minSubArrayLen(tar,inp))